pf/tfgen/main.go

```diff
@@ -60,6 +70,171 @@ func Main(provider string, info tfbridge.ProviderInfo) {
 	})
 }
 
+// Implements main() logic for a multi-provider build-time helper utility. By convention these utilities are
+// named pulumi-tfgen-$provider, for example when building a "random" provider the program would be called
+// pulumi-tfgen-random.
+//
+// The resulting binary is able to generate [Pulumi Package Schema] as well as provider SDK sources in various
+// programming languages supported by Pulumi such as TypeScript, Go, and Python.
+//
+// [Pulumi Package Schema]: https://www.pulumi.com/docs/guides/pulumi-packages/schema/
+func MainWithMuxer(provider string, infos ...tfbridge.Muxed) {
+	ctx := context.Background()
+
+	// It is safe to pass infos[0].GetInfo() to MainWithCusomGenerate because it is
+	// only used when passed to GenerateOptions.
+
+	gen := func(opts tfgen.GeneratorOptions) error {
+		// To avoid double-initializing the sink (and thus panicking), we eagerly
+		// initialize here.
+		if opts.Sink == nil {
+			diagOpts := diag.FormatOptions{
+				Color: cmdutil.GetGlobalColorization(),
+				Debug: opts.Debug,
+			}
+			cmdutil.InitDiag(diagOpts)
+			opts.Sink = cmdutil.Diag()
+		}
+
+		muxedInfo := infos[0].GetInfo()
+		schemas := []schema.PackageSpec{}
+		var errs multierror.Error
+
+		for i, union := range infos {
+			// Concurrency safety
+			i := i
+			union := union
+
+			anErr := func(err error) bool {
+				if err == nil {
+					return false
+				}
+				errs.Errors = append(errs.Errors, fmt.Errorf("Muxed proider[%d]: %w", i, err))
+				return true
+			}
+
+			info := union.SDK
+			if pfInfo := union.PF; pfInfo != nil {
+				if err := notSupported(opts.Sink, pfInfo.ProviderInfo); anErr(err) {
+					continue
+				}
+
+				shimInfo := shimSchemaOnlyProviderInfo(ctx, *pfInfo)
+				info = &shimInfo
+			}
+
+			// Initially, muxedInfo **is** infos[0]. We don't want to perform
+			// any destructive edits on already defined functions.
+			if i != 0 {
+				// Now we purge already defined resources from the next schema,
+				// marking the all remaining resources as defined for future
+				// schemas.
+				muxedInfo.IgnoreMappings = append(muxedInfo.IgnoreMappings,
+					info.IgnoreMappings...)
+
+				rDefined := func(k string) bool {
+					_, defined := muxedInfo.Resources[k]
+					if defined {
+						return true
+					}
+					_, defined = muxedInfo.ExtraResources[k]
+					return defined
+				}
+
+				fDefined := func(k string) bool {
+					_, defined := muxedInfo.DataSources[k]
+					if defined {
+						return true
+					}
+					_, defined = muxedInfo.ExtraFunctions[k]
+					return defined
+				}
+
+				for k, v := range info.Resources {
```

Comment:
This code is mixing ProviderInfo structures?

Reply:
Yes. This is how we recover standard "unmapped resource/datasource" error messages.